src/WebJobs.Extensions.CosmosDB/Trigger/CosmosDBTriggerListener.cs

```diff
@@ -87,10 +87,10 @@ public async Task StopAsync(CancellationToken cancellationToken)
         {
             try
             {
-                if (this._host != null && this._listenerStarted)
+                if (this._host != null)
                 {
                     await this._host.UnregisterObserversAsync();
-                    this._listenerStarted = false;
+                    Interlocked.CompareExchange(ref this._listenerStatus, ListenerNotRegistered, ListenerRegistered);
```

Comment:
There's a time when you actually will set it to NotRegistered while it is Registering (during the call to Cancel). But b/c this is a single operation, you should remove the Interlocked.CompareExchange here. Just set it straight to NotRegistered, no matter what.